File: Book/C/Fig_7.4_schroProg.c
// sample compile command: "gcc -fopenmp -c Fig_7.4_schroProg.c" to generate *.o object file

#include <stdbool.h>
#include <omp.h>
#include <stdio.h>
// Three functions we use but do not show here:
// 1. Set seed for a pseudo-random sequence
void seedIt(long *val);
// 2. function to flip a coin (randomly return true or false)
bool flip(long *coin);
// 3. wait a short random amount of time
double waitAbit();

int main()
{
   double wait_val;
   long rand, i; 
   int dead_or_alive;
   omp_set_num_threads(2);
   
   // "flip a coin" to choose which task is for the dead
   // cat and which for the living cat.
   long coin;
   seedIt(&coin);
   bool HorT = flip(&coin);
   
   printf("Schrodinger's program says the cat is");
    
   #pragma omp parallel shared(HorT, dead_or_alive)
   {
      // These tasks are participating in a data race        
      #pragma omp task
      {
         double val = waitAbit();
         dead_or_alive = HorT;
      }
      #pragma omp task
      {
         double val = waitAbit();
         dead_or_alive = !HorT;
      }
   } 
   if (dead_or_alive)
      printf(" alive. \n");
   else 
      printf(" dead. \n");

   return 0;
}

File: Book/Fortran/Fig_7.4_schroProg.f90
! sample compile command: "gfortran -fopenmp -c Fig_7.4_schroProg.f90" to generate *.o object file

program main
   use omp_lib
   implicit none
   external :: seedIt 
   logical, external :: flip
   real*8, external :: waitAbit

   real*8 :: wait_val, val
   integer*8 :: rand, i 
   logical :: dead_or_alive
   
   ! "flip a coin" to choose which task is for the dead
   ! cat and which for the living cat.
   integer*8 :: coin
   logical ::  HorT

   call omp_set_num_threads(2)
   call seedIt(coin)
   HorT = flip(coin)
   
   print *, "Schrodinger's program says the cat is"
    
   !$omp parallel shared(HorT, dead_or_alive) private(val)
      ! These tasks are participating in a data race        
      !$omp task
         val = waitAbit()
         dead_or_alive = HorT
      !$omp end task
      !$omp task
         val = waitAbit()
         dead_or_alive = .not. HorT
      !$omp end task
   !$omp end parallel
   if (dead_or_alive) then
      print *, " alive. "
   else 
      print *, " dead. "
   endif

end program main
